### PR TITLE
Prevent HMS from collecting stats when creating external tables

### PR DESCRIPTION
**Pull request by aalbu**
Fixes https://github.com/prestosql/presto/issues/3624.

This fix is attempting to cause Hive (2.x, 3.x) to skip computing file count stats (see https://github.com/apache/hive/blob/branch-2.3/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreUtils.java#L275-L280).

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -1061,6 +1061,9 @@ private static Table buildTableObject(
 
         if (external) {
             tableParameters.put("EXTERNAL", "TRUE");
+            // collecting stats (like number of files) at table creation time can cause the the DDL statement to time out
+            // when the table location contains a large number of files
+            tableParameters.put("DO_NOT_UPDATE_STATS", "true");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:1066]: Is it suitable to external tables only?

(it's likely, but please mention this in a comment or link to apache/hive repo, to the place where this is being `if`-d on)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 2]
```diff
@@ -1064,7 +1065,31 @@ public void alterTable(HiveIdentity identity, String databaseName, String tableN
                     .stopOnIllegalExceptions()
                     .run("alterTable", stats.getAlterTable().wrap(() -> {
                         try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
-                            client.alterTable(databaseName, tableName, table);
+                            Map<String, String> tableParameters = table.getParameters();
+                            String statsParam = "DO_NOT_UPDATE_STATS";
+                            if ("true".equals(tableParameters.get(statsParam))) {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1070]: Does it have to be conditional?

I don't think there is a situation where I would want metastore to collect the basic stats (we update them for managed tables, and for external tables you are now disabling them)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:1066]: Also, do we need to fill `tableParameters`, or only the `EnvironmentContext`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 3]
```diff
@@ -1064,7 +1065,31 @@ public void alterTable(HiveIdentity identity, String databaseName, String tableN
...
+                                EnvironmentContext context = new EnvironmentContext();
+                                context.setProperties(ImmutableMap.of(statsParam, "true"));
+                                // the DO_NOT_UPDATE_STATS is not supposed to be a persistent parameter
+                                table.setParameters(tableParameters.entrySet().stream()
+                                        .filter(entry -> !statsParam.equals(entry.getKey()))
+                                        .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue)));
+                                alternativeCall(
+                                        () -> client,
+                                        ThriftHiveMetastore::defaultIsValidExceptionalResponse,
+                                        chosenGetTableAlternative,
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1080]:
```suggestion
                                        chosenAlterTableAlternative,
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 4]
```diff
@@ -1064,7 +1065,31 @@ public void alterTable(HiveIdentity identity, String databaseName, String tableN
...
+                                        metastoreClient -> {
+                                            metastoreClient.alterTableWithEnvironmentContext(databaseName, tableName, table, context);
+                                            return 0;
+                                        },
+                                        metastoreClient -> {
+                                            metastoreClient.alterTable(databaseName, tableName, table);
+                                            return 0;
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1087]: if return is required, use `return null` (`0` looks more meaningful)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (aalbu) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:1066]: > Is it suitable to external tables only?
Hive doesn't seem to make a distinction, but this you cannot create a managed table that points to a location with data.

> Also, do we need to fill tableParameters, or only the EnvironmentContext?
For version 3.x, providing the table parameter to Hive is sufficient.  Hive will skip collecting stats and will remove the parameters (see https://github.com/apache/hive/blob/branch-3.1/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java#L668-L684).  There is a comment in there that says "We want to avoid this property from being persistent".  In Presto, we use the table property as a way to communicate that the intent is to not collect stats.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java (hunk 3 above)
Comment (aalbu) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1080]: Copy pasta :)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java (hunk 2 above)
Comment (aalbu) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1070]: I was trying to minimize the number of code paths affected by this change, because I didn't fully understand potential consequences.

Now that you mention it, we don't need the stats collected by Hive anyway, so it would make sense to make this unconditional.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (aalbu) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:1066]: > Is it suitable to external tables only?

I tried to restrict to external table creation because managed tables cannot be created in a location that already contains data, so they're not affected by this problem.  Hive doesn't really make a distinction.

> Also, do we need to fill tableParameters, or only the EnvironmentContext?

I was initially using table parameters as a vehicle to inform `ThriftHiveMetastore` that it needed to make a different call.  But after our offline conversation where you pointed out that we needed to support Hive 1.x, I discovered that the context parameter is ignored in that version, however the table property is honored.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreClient.java [hunk 5]
```diff
@@ -71,6 +72,9 @@ void dropTable(String databaseName, String name, boolean deleteData)
     void alterTable(String databaseName, String tableName, Table newTable)
             throws TException;
 
+    void alterTableWithEnvironmentContext(String databaseName, String tableName, Table newTable, EnvironmentContext context)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreClient.java:75]: remove `void alterTable(String databaseName, String tableName, Table newTable)`, it's unused now

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 6]
```diff
@@ -126,10 +126,8 @@ public void testCreatePartitioned()
                 "WHERE n_nationkey <> 23", tableName));
 
         try {
-            if (getHiveVersionMajor() < 3) {
-                BasicStatistics tableStatistics = getBasicStatisticsForTable(onHive(), tableName);
-                assertThatStatisticsAreNotPresent(tableStatistics);
-            }
+            BasicStatistics tableStatistics = getBasicStatisticsForTable(onHive(), tableName);
+            assertThatStatisticsAreNotPresent(tableStatistics);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:130]: Add a comment why this assertion matters, so that someone doesn't put an `if` around this again (like i did).
E.g. link to the issue.

Comment (findepi) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:130]: Also, can you add a test for CREATE external table? 
(partitioned and not partitioned)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 7]
```diff
@@ -420,6 +501,24 @@ private static BasicStatistics basicStatisticsFromDescribeResult(QueryResult res
...
+                    .map(String::trim);
+
+            if (name.isPresent() && "Location:".equals(name.get())) {
+                return Optional.ofNullable(row.get(1))
+                        .map(Object::toString)
+                        .map(String::trim);
+            }
+        }
+        return Optional.empty();
+    }
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:520]: maybe:
https://github.com/prestosql/presto/blob/1b457a1188b44828d5aba7eb7e58d9aa45e1df28/presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java#L6469

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 8]
```diff
@@ -54,6 +55,41 @@ public void testCreateUnpartitioned()
...
 
+    @Test
+    public void testCreateExternalUnpartitioned()
+    {
+        String tableName = "test_basic_statistics_external_unpartitioned_presto";
+        String helperTableName = tableName + "_helper";
+
+        onPresto().executeQuery(format("DROP TABLE IF EXISTS %s", helperTableName));
+        onPresto().executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
+        onPresto().executeQuery(format("CREATE TABLE %s AS SELECT * FROM nation", helperTableName));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:66]: Do we need `helperTableName` table? can we use `nation` directly instead?

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 9]
```diff
@@ -144,6 +180,51 @@ public void testCreatePartitioned()
...
+                        "   n_regionkey bigint) " +
+                        "WITH (partitioned_by = ARRAY['n_regionkey'], external_location = '%s')",
+                        tableName,
+                        location.get()));
+            }
+            else {
+                fail("Could not determine location of managed table");
+            }
+            BasicStatistics statistics = getBasicStatisticsForTable(onHive(), tableName);
+            assertThatStatisticsAreZero(statistics);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:220]: Maybe in partitioned table case, we could call sync_partitions procedure and verify again.
(Unless it wouldn't work in that case, then keep as it is.)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 10]
```diff
@@ -184,43 +174,27 @@ public void testCreatePartitioned()
...
+                    "   n_name varchar(25), " +
+                    "   n_comment varchar(152), " +
+                    "   n_regionkey bigint) " +
+                    "WITH (partitioned_by = ARRAY['n_regionkey'], external_location = '%s')",
+                    tableName,
+                    tableLocation));
+            // in Hive 3, DESCRIBE FORMATTED returns 0 for table stats that are actually missing from the metastore
+            if (getHiveVersionMajor() < 3) {
+                BasicStatistics tableStatistics = getBasicStatisticsForTable(onHive(), tableName);
+                assertThatStatisticsAreNotPresent(tableStatistics);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:194]: else `assertThatStatisticsAreZero` ?

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 11]
```diff
@@ -501,22 +475,9 @@ private static BasicStatistics basicStatisticsFromDescribeResult(QueryResult res
...
-                    .map(String::trim);
-
-            if (name.isPresent() && "Location:".equals(name.get())) {
-                return Optional.ofNullable(row.get(1))
-                        .map(Object::toString)
-                        .map(String::trim);
-            }
-        }
-        return Optional.empty();
+        return onPresto().executeQuery("SELECT DISTINCT regexp_replace(\"$path\", '/[^/]*$', '') FROM " + tableName).<String>column(1).get(0);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:480]:
```suggestion
        return getOnlyElement(onPresto().executeQuery("SELECT DISTINCT regexp_replace(\"$path\", '/[^/]*$', '') FROM " + tableName).column(1));
```

-- fail if there are 2+ results

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 12]
```diff
@@ -947,7 +947,7 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
                     tableProperties.put(CSV_SEPARATOR_KEY, separator.toString());
                 });
 
-        // prevent HMS from gathering these stats at table creation, since it can be expensive
+        // set bogus table stats to prevent Hive 2.x from gathering these stats at table creation
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:950]:
```suggestion
        // Set bogus table stats to prevent Hive 2.x from gathering these stats at table creation.
        // These stats are not useful in itself and can take very long time to collect when creating an
        // external table over large data set.
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java [hunk 13]
```diff
@@ -1066,6 +1066,7 @@ public void alterTable(HiveIdentity identity, String databaseName, String tableN
                     .run("alterTable", stats.getAlterTable().wrap(() -> {
                         try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
                             EnvironmentContext context = new EnvironmentContext();
+                            // this prevents Hive 3.x from collecting table stats at table creation time
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java:1069]:
```suggestion
                            // This prevents Hive 3.x from collecting basic table stats at table creation time.
                            // These stats are not useful in itself and can take very long time to collect when creating an
                            // external table over large data set.
```

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 14]
```diff
@@ -144,6 +170,35 @@ public void testCreatePartitioned()
...
+    public void testCreateExternalPartitioned()
+    {
+        String tableName = "test_basic_statistics_external_partitioned_presto";
+
+        onPresto().executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
+
+        try {
+            String tableLocation = getTableLocation("nation");
+            onPresto().executeQuery(format("" +
+                    "CREATE TABLE %s (" +
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:183]: According to my intellij indentation is off here.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 15]
```diff
@@ -144,6 +171,33 @@ public void testCreatePartitioned()
...
+                "   n_name varchar(25), " +
+                "   n_comment varchar(152), " +
+                "   n_regionkey bigint) " +
+                "WITH (partitioned_by = ARRAY['n_regionkey'], external_location = '%s')",
+                tableName,
+                tableLocation));
+        try {
+            assertThatTableHasNoStatsFlexible(tableName);
+            onPresto().executeQuery(format("CALL system.sync_partition_metadata('default', '%s', 'FULL')", tableName));
+            assertThatTableHasNoStatsFlexible(tableName);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:194]: add an assertion for SHOW STATS on presto side (table must not appear as empty)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 16]
```diff
@@ -191,15 +204,24 @@ public void testCreateExternalPartitioned()
...
-                        assertThat(rowCounts.get(rowCounts.size() - 1)).isNotNull().isNotEqualTo(0);
-                    });
+            Optional<Integer> rowCountIndex = result.tryFindColumnIndex("row_count");
+            if (rowCountIndex.isPresent()) {
+                List<Object> rowCounts = result.column(rowCountIndex.get());
+                // the last row has table stats
+                assertThat((Double) rowCounts.get(rowCounts.size() - 1)).isNotNull().isGreaterThan(0.0);
+            }
+            else {
+                fail("Unable to find column row_count in ANALYZE result");
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:223]: int rowCountIndex = result.tryFindColumnIndex("row_count")
  .orElseThrow

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 17]
```diff
@@ -191,15 +204,24 @@ public void testCreateExternalPartitioned()
...
             assertThatTableHasNoStatsFlexible(tableName);
             onPresto().executeQuery(format("CALL system.sync_partition_metadata('default', '%s', 'FULL')", tableName));
-            assertThatTableHasNoStatsFlexible(tableName);
+            // in Hive 3, syncing partitions causes stats to be populated
+            if (getHiveVersionMajor() < 3) {
+                BasicStatistics tableStatistics = getBasicStatisticsForTable(onHive(), tableName);
+                assertThatStatisticsAreNotPresent(tableStatistics);
+            }
+
+            onPresto().executeQuery(format("ANALYZE %s", tableName));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:213]: remove?

We're supposedly testing CREATE TABLE that it doesn't produce bogus stats.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 18]
```diff
@@ -54,6 +55,32 @@ public void testCreateUnpartitioned()
...
+
+        try {
+            String location = getTableLocation("nation", false);
+            onPresto().executeQuery(format("" +
+                    "CREATE TABLE %s (" +
+                    "   n_nationkey bigint, " +
+                    "   n_regionkey bigint, " +
+                    "   n_name varchar(25), " +
+                    "   n_comment varchar(152)) " +
+                    "WITH (external_location = '%s')",
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:73]: `nation` table is `format = 'TEXTFILE'`, so use this here too

(you can verify the table format with `select count(*) from the table`)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 19]
```diff
@@ -54,6 +55,32 @@ public void testCreateUnpartitioned()
...
+                    tableName,
+                    location));
+            BasicStatistics statistics = getBasicStatisticsForTable(onHive(), tableName);
+            assertThatStatisticsAreZero(statistics);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:77]: This is wrong. i checked with `show stats for test_basic_statistics_external_unpartitioned_presto;` that this indeed reports row count 0, but the table is not empty. It should report NULL

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java [hunk 20]
```diff
@@ -144,6 +171,49 @@ public void testCreatePartitioned()
         }
     }
 
+    @Test
+    public void testCreateExternalPartitioned()
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:175]: Since we're doing a lot of back and forth to satisfy every supported metastore version, let's move partitioned tables to a separate PR and focus on unpartitioned for now.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java (hunk 19 above)
Comment (findepi) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:77]: I guess this is actually a pre-existing thing, is it?

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java (hunk 18 above)
Comment (findepi) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBasicTableStatistics.java:73]: `textfile_field_separator = '|'` is important too

**Comment by findepi** [comment]
Merged, thanks!
